File: geeksgforgeeks/challenge10_lps/LPS.java
/*
    Created by Avinash on 9th November, 2017
*/
import java.io.*;
import java.util.*;
public class LPS
{
	public static int output;
	public static void main(String[] args)
	{
		BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
		String data;
		int lps = 0;
		try
		{
		System.out.println("Enter the string : ");
		data = read.readLine();
		Helper link = new Helper();	
		lps = link.findLPSDP(data.toCharArray(),0,data.length()-1);
		}
		catch(IOException e)
		{
			e.printStackTrace();
		}
		System.out.print("The length of longest palindromic sequence is "+lps);
	}
}

class Helper
{
	public int findLPSRec(char[] data, int i, int j)
	{
		if( i == j)
			return 1;
		if(data[i] == data[j] && i + 1 == j)
			return 2;
		if(data[i] == data[j])
			return 2 + findLPSRec(data, i + 1, j - 1);
		return Math.max(findLPSRec(data, i, j - 1), findLPSRec(data, i + 1, j));	
	}

	public int findLPSDP(char[] data, int x, int y)
	 {
	 	int n = data.length;
	 	int cache[][] = new int[n][n];
	 	for(int i = 0; i < n; i++) // for all subsequences of length 1, LPS is 1
	 		cache[i][i] = 1;
	 	for(int sub = 2; sub <= n; sub++) //We now check for sequences of length 2,3,4..n
	 	{
	 		for(int i = 0; i < n - sub + 1; i++)
	 		{
	 			int j = i + sub - 1;
	 			if(data[i] == data[j] && sub == 2)
	 				cache[i][j] = 2;
	 			else if(data[i] == data[j])
	 				cache[i][j] = 2 + cache[i+1][j-1];
	 			else
	 				cache[i][j] = Math.max(cache[i+1][j], cache[i][j-1]);
	 		}
	 	}
	 	return cache[0][n-1];		
	}
  	
}	                    	
			

File: geeksgforgeeks/challenge10_lps/lps.py
# Created by Avinash on 10th Nov, 2017
import sys
def process():
	data = input("Enter the string : ")
	lps = lpsCache(data,0,len(data)-1)
	print("The length of longest palindromic sequence is %-5d" %lps)

def lpsrec(data,i,j):
	if i == j :
		return 1
	if data[i] == data[j] and i + 1 == j:
		return 2
	if data[i] == data[j] :
		return 2 + lpsrec(data, i + 1, j - 1)
	return max(lpsrec(data,i,j-1), lpsrec(data,i+1,j))	

def lpsCache(data,i,j):
	n = len(data);
	cache = [[ 0 for j in range(n)] for i in range(n)]
	for i in range(n):
		cache[i][i] = 1
	for sub in range(2,n+1):	
		for i in range(0,n-sub+1):
	 		j = i + sub - 1
	 		if data[i] == data[j] and sub == 2 :
	 			cache[i][j] = 2 
	 		elif data[i] == data[j] :
	 			cache[i][j] = 2 + cache[i+1][j-1]
	 		else :
	 			cache[i][j] = max(cache[i+1][j], cache[i][j-1])
	 	
	return cache[0][n-1]
					

if __name__=="__main__":
	process()
		
			